tls/datadog_checks/tls/tls.py

```diff
@@ -0,0 +1,348 @@
+
+        with closing(sock):
+            try:
+                with closing(self.tls_context.wrap_socket(sock, server_hostname=self._server_hostname)) as secure_sock:
+                    der_cert = secure_sock.getpeercert(binary_form=True)
+                    protocol_version = secure_sock.version()
+            except Exception as e:
+                # https://docs.python.org/3/library/ssl.html#ssl.SSLCertVerificationError
+                err_code = getattr(e, 'verify_code', None)
+                message = getattr(e, 'verify_message', str(e))
+                self.service_check(self.SERVICE_CHECK_VALIDATION, self.CRITICAL, tags=self._tags, message=message)
+
+                # There's no sane way to tell it to not validate just the expiration
+                # This only works on Python 3.7+, see: https://bugs.python.org/issue28182
+                # https://github.com/openssl/openssl/blob/0b45d8eec051fd9816b6bf46a975fa461ffc983d/include/openssl/x509_vfy.h#L109
+                if err_code == 10:
+                    self.service_check(
+                        self.SERVICE_CHECK_EXPIRATION, self.CRITICAL, tags=self._tags, message='Certificate has expired'
+                    )
+
+                return
+
+        try:
+            cert = load_der_x509_certificate(der_cert, default_backend())
+        except Exception as e:
+            self.service_check(
+                self.SERVICE_CHECK_VALIDATION,
+                self.CRITICAL,
+                tags=self._tags,
+                message='Unable to parse the certificate: {}'.format(e),
+            )
+            return
+
+        self.check_protocol_version(protocol_version)
+        self.validate_certificate(cert)
+        self.check_age(cert)
+
+    def check_local(self, instance):
+        if not self._local_cert_path:
+            raise ConfigurationError('You must specify `local_cert_path` in your configuration file.')
+        elif self._validate_hostname and not self._server_hostname:
+            raise ConfigurationError(
+                'You must specify `server_hostname` in your configuration file, or disable `validate_hostname`.'
+            )
+
+        try:
+            with open(self._local_cert_path, 'rb') as f:
+                cert = f.read()
+        except Exception as e:
+            self.service_check(
+                self.SERVICE_CHECK_VALIDATION,
+                self.CRITICAL,
+                tags=self._tags,
+                message='Unable to open the certificate: {}'.format(e),
+            )
+            return
+
+        try:
+            cert = self.local_cert_loader(cert)
+        except Exception as e:
+            self.service_check(
+                self.SERVICE_CHECK_VALIDATION,
+                self.CRITICAL,
+                tags=self._tags,
+                message='Unable to parse the certificate: {}'.format(e),
+            )
+            return
+
+        self.validate_certificate(cert)
```

Comment:
Most probably a stupid question but why aren't we checking the protocol version here?

Reply:
No connection, just local file